x-pack/platform/plugins/shared/alerting/server/maintenance_window_events/task.ts

```diff
@@ -0,0 +1,284 @@
+            'lte',
+            endRangeDate
+          );
+
+          const statusFilter = getStatusFilter();
+
+          const filter = nodeBuilder.and([startRangeFilter, endRangeFilter, statusFilter]);
+
+          const findMaintenanceWindowsResult = await findMaintenanceWindowsSo({
+            filter,
+            savedObjectsClient,
+            logger,
+          });
+
+          totalMaintenanceWindowsWithGeneratedEvents =
+            await generateEventsAndUpdateMaintenanceWindowSavedObjects({
+              logger,
+              savedObjectsClient,
+              maintenanceWindowsSO: findMaintenanceWindowsResult,
+              startRangeDate,
+            });
+
+          const updatedState: LatestTaskStateSchema = {
+            runs: state.runs + 1,
+            total_mw_with_generated_events: totalMaintenanceWindowsWithGeneratedEvents,
+          };
+          logger.info(
+            `Maintenance windows events generator task updated ${updatedState.total_mw_with_generated_events} maintenance windows successfully`
+          );
+
+          return {
+            state: updatedState,
+            schedule: SCHEDULE,
+          };
+        } catch (e) {
+          logger.warn(`Error executing maintenance windows events generator task: ${e.message}`);
+          const updatedState: LatestTaskStateSchema = {
+            runs: state.runs + 1,
+            total_mw_with_generated_events: totalMaintenanceWindowsWithGeneratedEvents,
+          };
+          return {
+            state: updatedState,
+            schedule: SCHEDULE,
+          };
+        }
+      },
+    };
+  };
+}
+
+export function getStatusFilter() {
+  const running = nodeBuilder.is('maintenance-window.attributes.events', 'now');
+  const upcoming = nodeBuilder.and([
+    nodeTypes.function.buildNode(
+      'not',
+      nodeBuilder.is('maintenance-window.attributes.events', 'now')
+    ),
+    nodeBuilder.range('maintenance-window.attributes.events', 'gte', 'now'),
+  ]);
+  const finished = nodeBuilder.and([
+    nodeTypes.function.buildNode(
+      'not',
+      nodeBuilder.range('maintenance-window.attributes.events', 'gte', 'now')
+    ),
+    nodeBuilder.range('maintenance-window.attributes.expirationDate', 'gte', 'now'),
+  ]);
+
+  return nodeBuilder.or([running, upcoming, finished]);
+}
+
+export async function findMaintenanceWindowsSo({
+  savedObjectsClient,
+  logger,
+  filter,
+}: {
+  logger: Logger;
+  savedObjectsClient: ISavedObjectsRepository;
+  filter: KueryNode;
+}) {
+  try {
+    const finder = await savedObjectsClient.createPointInTimeFinder<MaintenanceWindowAttributes>({
+      type: MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE,
+      namespaces: ['*'],
+      perPage: 1000,
+      sortField: 'updatedAt',
+      sortOrder: 'desc',
+      filter,
+    });
+
+    let result: Array<SavedObject<MaintenanceWindowAttributes>> = [];
+    for await (const findResults of finder.find()) {
+      result = result.concat(findResults.saved_objects);
```

Comment:
By doing a `concat`, we load all the SOs in memory. We should batch process the SOs and update them on each loop iteration. The flow could look like (pseudo code)

```
// in eventsGeneratorTaskRunner.ts

await updateMaintenanceWindowsEvents(...)

// below in the same file

const updateMaintenanceWindowsEvents = (...) => {

const soFinder = getSOFinder(...)

for await (const res of soFinder.find()) {  }
  totalEvents =+ generateEventsAndUpdateMaintenanceWindowSavedObjects(res.saved_objects)
}
```

Reply:
Updated the flow in https://github.com/elastic/kibana/pull/219261/commits/b039796cd98d269349a37d5ef43cae2a01b02b59